### PR TITLE
[discover-next][bug] add back data set navigator to control state

### PR DESCRIPTION
**Pull request by kavilla**
### Description

Re-add the data set navigator upon other changes with passing tests.

### Issues Resolved

n/a

## Changelog

- skip

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by joshuali925** [comment]
i have another PR for discover-2.0-1 branch, should i rebase and change base to main? https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7494

**Comment by abbyhu2000** [comment]
Pulled down the branch and ran, there is too much space between the code editor and saved query management when the navigation sidebar is collapsed, will fix that in my next PR.

Screenshot 2024-07-26 at 3 49 06 PM

**Review by abbyhu2000 — DISMISSED** [review]

File: src/plugins/data/public/antlr/opensearch_sql/code_completion.ts [hunk 1]
```diff
@@ -86,7 +85,7 @@ export const getSuggestions = async ({
           tableNames,
           suggestions.suggestValuesForColumn as string,
           api,
-          connectionService
+          services
```
Comment (paulstn) [inline src/plugins/data/public/antlr/opensearch_sql/code_completion.ts:88]: might need to remove these lines from here, since the other autocomplete prs were removed from this branch. this change was added in this commit, would only need to revert the changes done on this file: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7492/commits/bf057f2ac8468aba653b6ad2ea675413da1f1178#diff-f67c8b1b68c9ae8eeb8ea7d7dea38e59d32b7738dec7c9db5763cf74b96aa92fR46

File: src/plugins/discover/public/application/view_components/utils/use_index_pattern.ts [hunk 2]
```diff
@@ -25,16 +28,38 @@ import { getIndexPatternId } from '../../helpers/get_index_pattern_id';
...
+
+  const fetchIndexPatternDetails = useCallback(
+    async (id: string) => {
+      return await data.indexPatterns.get(id);
+    },
+    [data.indexPatterns]
+  );
+
+  useEffect(() => {
+    if (isQueryEnhancementEnabled) {
```
Comment (sejli) [inline src/plugins/discover/public/application/view_components/utils/use_index_pattern.ts:47]: Nit: single `if` here?
```
  if (
    isQueryEnhancementsEnabled &&
    dataset &&
    dataSet.type === SIMPLE_DATA_SET_TYPES.INDEX_PATTERN
  ) {
    fetchIndexPatternDetails(dataSet.id).then((ip) => setIndexPattern(ip));
  }

File: src/plugins/discover/public/application/view_components/utils/use_index_pattern.ts [hunk 3]
```diff
@@ -58,20 +83,31 @@ export const useIndexPattern = (services: DiscoverViewServices) => {
...
-        const newId = getIndexPatternId('', indexPatternList, config.get('defaultIndex'));
-        store!.dispatch(updateIndexPattern(newId));
-        fetchIndexPatternDetails(newId);
-      });
-    } else {
-      fetchIndexPatternDetails(indexPatternIdFromState);
+    if (!isQueryEnhancementEnabled) {
+      if (!indexPatternIdFromState) {
+        data.indexPatterns.getCache().then((indexPatternList) => {
+          const newId = getIndexPatternId('', indexPatternList, uiSettings.get('defaultIndex'));
```
Comment (sejli) [inline src/plugins/discover/public/application/view_components/utils/use_index_pattern.ts:89]: What if there's no `defaultIndex`?

**Review by sejli — APPROVED** [review]
Just some nits, think most of it is good.

File: src/plugins/query_enhancements/server/routes/data_source_connection/routes.ts [hunk 4]
```diff
@@ -18,7 +17,7 @@ export function registerDataSourceConnectionsRoutes(router: IRouter) {
     },
     async (context, request, response) => {
       const fields = ['id', 'title', 'auth.type'];
-      const resp = await context.core.savedObjects.client.find<DataSourceAttributes>({
+      const resp = await context.core.savedObjects.client.find({
```
Comment (ashwin-pc) [inline src/plugins/query_enhancements/server/routes/data_source_connection/routes.ts:20]: Why did we remove this type?

File: src/plugins/query_enhancements/public/services/connections_service.ts
Comment (ashwin-pc) [inline src/plugins/query_enhancements/public/services/connections_service.ts:1]: Do we still need this service? or is this the fast follow that @mengweieric was mentioning that we need to remove?

**Review by ashwin-pc — DISMISSED** [review]

**Review by ashwin-pc — APPROVED** [review]
Reapproving after adding the changeset

File: src/plugins/query_enhancements/public/services/connections_service.ts
Comment (mengweieric) [reply src/plugins/query_enhancements/public/services/connections_service.ts:1]: yea, this is the one I'm talking about, basically we need this PR but got reverted: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7463

**Review by abbyhu2000 — APPROVED** [review]

**Review by sejli — APPROVED** [review]
